---
set_title: !r test_title
output: 
  html_document:
    toc: yes
    toc_float:
      collapsed: true
    toc_depth: 3
    number_sections: false
    self_contained: yes
    css: corp-styles.css
    highlight: pygments
---

```{r, echo=FALSE}
htmltools::img(src = "logo.png",
               alt = 'logo',
               style = 'position:absolute; top:0; right:0; down:20; padding:5px; width:100px; height:auto;')
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/orfeo/cephfs/scratch/cdslab/shared/SCOUT/SPN01/races')
```

<!-- Libraries  -->

```{r libraries, echo=FALSE, warning=FALSE, include=FALSE, message = FALSE}

#load libraries
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
library(ggplot2)
library(patchwork)
library(rRACES)
library(dplyr)
library(tidyverse)
source("/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/report/plotting/signature_rRACES.R")
#source("/orfeo/LTS/LADE/LT_storage/lvaleriani/races/rRACES-examples/report/plotting/signature_rRACES.R")
source("/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/report/plotting/plot_genome_wide.R")
source("/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/report/plotting/dynamics_rRACES.R",local =T)
metadata <- read.table(file = "/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/report/SCOUT_metadata.csv",header = T,sep = "\t")
metadata <- read.table(file = "/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/report/SCOUT_metadata.csv",header = T,sep = "\t")
img_dir <- "/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/report/SCOUT_images/"
#simulation <- rRACES::recover_simulation("SPN01")
```

```{r params, include = FALSE}
#Load params from YAML if running outside R Markdown
if (!exists("params")) {
  params <- yaml::read_yaml("/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/SCOUT/SPN01/params_50x_0.3p.yml")
}
```

<!-- Load the data  -->

```{r laod_data, echo=FALSE, include=FALSE}
sample_forest <- load_samples_forest(params$files$sample_forest)
phylo_forest <- load_phylogenetic_forest(params$files$phylo_forest)
files_cna <- list.files(params$cna_dir, full.names = T)
sample_names <- sapply(files_cna, function(path) {
  base_name <- basename(path)
  sub("_cna.rds$", "", base_name)
})
number_of_samples <- length(sample_names)
cna_data <- lapply(list.files(params$cna_dir,recursive = T,full.names = T),readRDS)
```

---
title: "Report for `r params$spn`"
---

```{r get_spn_info, include = FALSE}
#Load params from YAML if running outside R Markdown
info_spn <- metadata %>% filter(SPN_ID==params$spn)
```

This report shows the results for patient **`r params$spn`** of SCOUT cohort. `r info_spn$Description` **`r params$spn`** is profiled by `r info_spn$N_Samples` samples collected at `r info_spn$N_Timepoints` time-point.

<!-- ```{r pressure, echo=FALSE, out.width = '100%'} -->

<!-- spn_image <- paste0(img_dir,params$spn,".png") -->

<!-- knitr::include_graphics(spn_image) -->

<!-- ``` -->

# Tissue Simulation and Sampling

This part of the report shows how the tumour growth in time. Muller plot and time series plot show changing in cellular growth of the different clones.

## Tissue dynamics

`r params$spn` growth dynamics is summarized by the muller plot toghter with a temporal line showing the time in which sampling is performed.

```{r grouth_dynamics, out.width = '100%',fig.align='center',fig.height=8}
snapshot_dir <- params$files$sim
# knitr::opts_knit$set(root.dir = snapshot_dir)
snapshot_name <- params$spn

# simulation <- rRACES::recover_simulation("SPN01")
plot <- plot_tumour_dynamics(paste0(snapshot_dir),sample_forest)
plot$plot_dynamics
```

## Sampling

`r params$spn` is sampled at `r info_spn$N_Timepoints` time-point, where a total of `r info_spn$N_Samples` samples are derived. Sampling dynamics is reported in plot below.

```{r samlping_dynamics, echo=FALSE, out.width = '100%',fig.align='center',fig.height=3, fid.width=10}
print(plot$plot_sampling)
```

General information about the samples are reported in the table below.

```{r test_table}
library(reactable)
library(sparkline)
source("/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/report/plotting/tables.R", local = knitr::knit_global())
table <- samples_table(snapshot=params$files$sim,
                       forest=params$files$sample_forest)

reactable(table,filterable = FALSE, resizable = T,highlight = T,defaultPageSize = 10,
          bordered = TRUE, striped = TRUE)
```

## Phylogenetic forest

Sample forest plot with annotated MRCAs for all available sample is showed below. Cells are colored accordining to clone membership.

```{r forest, echo=FALSE,fig.height=6, out.width = '100%'}
color_map_clones <-  get_clone_map(sample_forest)
plt_sample <- plot_forest(sample_forest, 
                          color_map = color_map_clones) +
  theme(legend.box = "vertical", 
        legend.box.just = "left",
        legend.spacing = unit(0.01, "cm"))
annotate_forest(plt_sample, phylo_forest, samples = TRUE, MRCAs = TRUE, driver = F)
```

# Tumour evolutionary parameters

Mutational processes are added to the sampled cells based on the tumour type. In particular, the mutation engine is set up by specifing:

1.  Mutation rates of SNVs, INDELs and CNAs;
2.  Driver SNVs and CNAs;
3.  SBS and ID mutational signatures.

## Mutation rate

The table below shows the mutation rate for SNVs, INDELs and CNAs added during `MutationEngine` set up.

```{r species_info}
library(reactable)
library(sparkline)
species_table <- phylo_forest$get_species_info()
reactable(species_table,filterable = FALSE, resizable = T,highlight = T,defaultPageSize = 10,
          bordered = TRUE, striped = TRUE, fullWidth = TRUE)

```

## Drivers event

For `r params$spn` drivers SNVs and CNAs are added to a particular clone, each chatacterized by a specific mutation rate (see above). The following table summarizes the genomic and evolutionary features of the clones.

The somatic mutation drivers (SNVs and INDELs) are:

```{r driver_event}
library(reactable)
library(sparkline)
driver_table <- phylo_forest$get_driver_mutations()

mut_event <- driver_table %>% 
  filter(type != 'CNA') %>% 
  filter(type != 'WGD') %>%
  select(-CNA_type, -order, -allele, -src_allele)

reactable(mut_event, filterable = FALSE, resizable = T,highlight = T,defaultPageSize = 5,
          bordered = TRUE, striped = TRUE, fullWidth = FALSE)
```

And the copy number alteration (CNA) driver events are:

```{r cna driver}
get_event <- function(driver){
  event <- driver %>% 
    filter(type == 'CNA') %>% 
    select(-ref, -alt) %>% 
    group_by(mutant, chr, start, end, type) %>% 
    summarize(A0 = sum(CNA_type == 'A' & src_allele == '0'),
              A1 = sum(CNA_type == 'A' & src_allele == '1'),
              D0 = sum(CNA_type == 'D' & allele == '0'),
              D1 = sum(CNA_type == 'D' & allele == '1')) %>% 
    mutate(M = 1 + A0 - D0,
           m = 1 + A1 - D1) %>% 
    mutate(major = max(M, m),
           minor = min(M, m)) %>%
    mutate(CN = paste(major, minor, sep = ':'),
           length = end - start) %>% 
    select(mutant, type, chr,  start, end, length, major, minor, CN) 
  
  return(event)
}

wgd <- driver_table %>% filter(type == 'WGD') %>% select(mutant, type, chr, start, end)
cna_event <- get_event(driver_table)
cna_event <- bind_rows(cna_event, wgd)


reactable(cna_event, filterable = FALSE, resizable = T,highlight = T,defaultPageSize = 5,
          bordered = TRUE, striped = TRUE, fullWidth = FALSE)
```

## Copy number events

Copy number profiles (passenger and driver) have specific segmentation given by the alterations that are appearing during tumour evolution. Segmented data are mapped to know karyotypes. The alluvial plot below shows how CN events are distributed between different samples:

```{r read_cna_data, echo=FALSE,fig.height=3,out.width = '100%'}
alluvial_plot <- alluvial_plot_karyotypes(files_cna)
alluvial_plot
```

Th single cell heatmap of copy number profiles across different clones and samples can be downloaded from this link: [single_cell_heatmap](/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/report/heatmap.pdf).

### Genomic Karyotype Composition Across Samples and Subclones  

This plot illustrates the percentage of the genome exhibiting a specific karyotype for each sample and its corresponding subclones. By quantifying the genomic regions affected by distinct karyotypic alterations, the visualization provides insights into the structural variations present within each sample. Comparing karyotype distributions across subclones helps to reveal intra-tumor heterogeneity and the evolutionary dynamics shaping the genomic landscape of the tumor.

```{r stats_cna, echo=FALSE,fig.height=3,out.width = '100%'}
plot_stats_cn(files_cna)
```

This circos plot provides a comprehensive overview of the segmented genomic data across all samples and their respective subclonal compositions. Each segment is color-coded based on its karyotype, allowing for a clear visualization of structural variations across the genome. Additionally, the transparency of the colors represents the cancer cell fraction (CCF), highlighting the prevalence of each segment within the tumor population. This integrated visualization captures the complexity of genomic alterations, enabling the exploration of karyotypic diversity and clonal architecture within and across samples.

```{r seg_cna, echo=FALSE,fig.height=3,out.width = '100%'}
plot_circle_segments(files_cna)
```

## Signature exposure

Mutational signatures provide insights into the underlying mutational processes shaping tumor evolution. These signatures are characteristic patterns of single base substitutions (SBS) and small insertions and deletions (IDs) that arise from distinct biological mechanisms, such as DNA damage, repair deficiencies, or external exposures. In this analysis, we simulate both SBS and ID signatures, capturing their evolution across time.

The selection of mutational signatures is based on the tumor type of interest, ensuring that the simulated processes reflect those observed in real-world cancer genomes. By modeling these signatures, we aim to understand how different mutational processes contribute to tumor progression and how they evolve over time.

For `r params$spn` the following SBS and ID signatures are added to the simulated tumour:

```{r mutation_engine, echo=FALSE}
exposure_table <- phylo_forest$get_exposures()
reactable(exposure_table,filterable = FALSE, resizable = T,highlight = T,defaultPageSize = 10,
          bordered = TRUE, striped = TRUE, fullWidth = TRUE)
```

### Phylogenetic tree mapping

Mutational signature exposures are mapped onto the phylogenetic tree, illustrating how their contributions change along the evolutionary trajectory of the tumor. This representation helps identify when specific mutational processes emerge, expand, or decline over time.

```{r forest_exposure, echo=FALSE, fig.width=9, out.width = '100%'}
plot_signatures <- plot_exposure_evolution(sample_forest, phylo_forest, snapshot_dir)
print(plot_signatures$sign_tree)
```

### Muller Plot Representation
The muller plot depicts the changing proportions of mutational signatures over time, capturing shifts in dominant mutational processes across different clonal and subclonal populations. This visualization highlights how signature exposures evolve dynamically during tumor development.

```{r muller_exposure, echo=FALSE, out.width = '100%',fig.align='center'}
print(plot_signatures$sign_muller)
```

# Sequencing simulation

Different combinations of purity and coverage have been simulate with parlleilze approach. You can find the report showing the results of specific purity-coverage combinations in the links listed in thefollowing table:

```{r table_with_seq_report, }
purities <- c("03","06","09")
coverages <- c(50,100,150,200)
df <- expand.grid(purity = purities, coverage = coverages, stringsAsFactors = FALSE)

# Add a column for path_to_report (modify as needed)

df$path_to_report <- paste0("[report](","Report_Seq_",df$coverage,"x_",df$purity,"p.html)")
  # paste0("reports/purity_", df$purity, "p_coverage_", df$coverage, "x.html")
reactable(df,filterable = FALSE, resizable = T,highlight = T,defaultPageSize = 12,
          bordered = TRUE, striped = TRUE, fullWidth = TRUE)
```
