---
set_title: !r test_title
output: 
  html_document:
    toc: yes
    toc_float:
      collapsed: true
    toc_depth: 3
    number_sections: true
    self_contained: yes
    css: corp-styles.css
    highlight: pygments
---

```{r, echo=FALSE}
htmltools::img(src = "logo.png",
               alt = 'logo',
               style = 'position:absolute; top:0; right:0; down:20; padding:5px; width:100px; height:auto;')
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<!-- Libraries  -->

```{r libraries, echo=FALSE, warning=FALSE, include=FALSE, message = FALSE}

#load libraries
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
library(ggplot2)
library(patchwork)
library(rRACES)
library(dplyr)
library(tidyverse)
source("plotting/utils.R")
source("plotting/plot_genome_wide.R")
metadata <- read.table(file = "SCOUT_metadata.csv",header = T,sep = "\t")
img_dir <- "SCOUT_images/"
```

```{r params, include = FALSE}
#Load params from YAML if running outside R Markdown
if (!exists("params")) {
  params <- yaml::read_yaml("/orfeo/cephfs/scratch/cdslab/ggandolfi/prj_races/REPO_UPDATED/rRACES-examples/SCOUT/SPN01/params.yml")
}
```

<!-- Load the data -->

```{r load_data, include = FALSE, echo =TRUE}
coverage <- params$sequencing$coverage
purity <- params$sequencing$purity
seq_dir <- params$files$seq_res
phylo_forest <- load_phylogenetic_forest(params$files$phylo_forest)
seq_res_tumour <- readRDS(paste0(seq_dir,"purity_",purity,"/seq_results_muts_merged_coverage_",coverage,"x.rds"))
seq_res_normal_rds <- paste0(seq_dir,"seq_results_muts_merged_coverage_30x.rds")
seq_res_normal <- readRDS(seq_res_normal_rds)
files_cna <- list.files(params$cna_dir, full.names = T)
print(files_cna)
sample_names <- sapply(files_cna, function(path) {
  base_name <- basename(path)
  sub("_cna.rds$", "", base_name)
})
cna_data <- lapply(list.files(params$cna_dir,recursive = T,full.names = T),readRDS)
```


# General sequencing information

Tumor samples have been sequenced with a coverage of `r params$sequencing$coverage`x and a simulated purity of `r params$sequencing$purity`. The normal sample is simulated from a real individual at a sequencing depth of 30x.

```{r individ}
library(reactable)
library(sparkline)

germ <- phylo_forest$get_germline_subject()
reactable(germ,filterable = FALSE, resizable = T,highlight = T,defaultPageSize = 5,
          bordered = TRUE, striped = TRUE)
```

```{r filter_muts, include = FALSE}
normal_mutations_filtered <- seq_res_normal %>%
  filter(normal_sample.VAF < 0.8 & normal_sample.VAF >0.2)
```

## Normal sample
```{r general_sequencing_normal, echo=FALSE,out.width = '80%',fig.align='center',fig.height=2}
dp <- seq_res_normal %>%
  seq_to_long() %>%
  ggplot(aes(x=DP))+
  ggplot2::geom_histogram()+
  geom_vline(aes(xintercept = mean(DP)),color = 'black',
                 linetype = 'dashed',
                  size = .5)+
  geom_vline(aes(xintercept = 30),colour = "red") +
  my_ggplot_theme()+
  scale_color_manual(name = "statistics", values = c("real mean" = "black", "expected mean" = "red"))
dp
```

## Tumour samples

```{r general_sequencing_tumour, echo=FALSE,out.width = '80%',fig.align='center',fig.height=2}
dp <- seq_res_tumour %>%
  seq_to_long() %>%
  ggplot(aes(x=DP))+
  ggplot2::geom_histogram()+
  geom_vline(aes(xintercept = mean(DP)),
                 color = 'black',
                 linetype = 'dashed',
                  size = .5)+
  geom_vline(aes(xintercept = coverage),colour = "red") +
  facet_wrap(~sample_name)+
  my_ggplot_theme()
dp
```

# Genome wide plots

## Depth Ratio

```{r depth_ratio, echo=FALSE,out.width = '100%',fig.align='center',fig.height=6}
g_seq <- left_join(normal_mutations_filtered,seq_res_tumour,
                   by =c("chr","chr_pos", "ref", "alt","causes","classes")) %>%
      as.data.frame()

dr_plots <- lapply(sample_names, function(s){
      dr <- plot_DR_n(seq_res = g_seq,sample = s)
})
wrap_plots(dr_plots,ncol=1)
```

## BAF

```{r baf, echo=FALSE,out.width = '100%',fig.align='center',fig.height=6}

baf_plots <- lapply(sample_names, function(s){
      baf <- plot_BAF_n(seq_res = g_seq,sample = s)
})
wrap_plots(baf_plots,ncol=1)
```

## VAF

```{r vaf, echo=FALSE,out.width = '100%',fig.align='center',fig.height=6}
s_seq <- seq_res_tumour %>%
  filter(classes!="germinal")
vaf_plots <- lapply(sample_names, function(s){
      vaf <- plot_VAF_n(seq_res = s_seq,sample = s)
})
wrap_plots(vaf_plots,ncol=1)
```


## Karyotypes information

```{r read_cna_data, echo=FALSE,fig.height=3,out.width = '100%'}
alluvial_plot <- alluvial_plot_karyotypes(files_cna)
alluvial_plot
```

## Segmented Data

```{r segmented_data, echo=FALSE,out.width = '100%'}
segmented_plot <- plot_clone_segments(files_cna)
segmented_plot
```
