---
title: "ProCESS Copy Number"
output: html_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

options(crayon.enabled = F)
```

```{r setup, warning=FALSE, message=FALSE}
set.seed(1234567)
rm(list = ls())
library(ProCESS)
require(dplyr)
library(knitr)
library(ggplot2)
library(patchwork)
```

## Copy Number

Let's see how one can produce a copy number event using ProCESS.
We will try to model a CNLOH.
To do so we need to create three populations: 

- A will acquire the "clonal mutations" for the subsequent populations
- B will be A's child and will acquire a drive CNLOH
- C will be B's childr

As always we need to follow some steps:

- Create the populations and sample cells
- Build mutation engine and define driver CN over the cells
- Simulate the sequencing and plot the results

## Sample cell populations

As already said, let's create three populations with same evolutionary parameters with the following hierarchy: A -> B -> C

```{r, message=FALSE, warning=FALSE}
sim <- new(Simulation, "Border Growth")

# Set the "border" growth model
sim$duplicate_internal_cells <- TRUE

# Set the death activation level to avoid drift
sim$death_activation_level <- 50

# Add mutants
sim$update_tissue(100, 100)
sim$add_mutant(name = "A", growth_rates = 0.1, death_rates = 0)
sim$place_cell("A", 50, 50)
sim$run_up_to_time(500)

n_w <- n_h <- 40
ncells <- 0.9*n_w*n_h
bbox <- sim$search_sample(c("A" = ncells), n_w, n_h)
sim$sample_cells("SA", bbox$lower_corner, bbox$upper_corner)
plot_tissue(sim, num_of_bins = 100)

# Add "B" mutant
sim$add_mutant(name="B", growth_rates = 0.5, death_rates=.01)
sim$mutate_progeny(sim$choose_cell_in("A"), "B")
sim$update_rates("A", c(growth = 0, death=.01))
sim$run_up_to_size("B", 100*100*.9)
plot_tissue(sim, num_of_bins = 100)

# Add "C" mutant
sim$add_mutant(name="C", growth_rates = 0.5, death_rates=.01)
sim$mutate_progeny(sim$choose_cell_in("B"), "C")
sim$update_rates("B", c(growth = 0, death=.01))
sim$run_up_to_size("C", 9000)

plot_tissue(sim, num_of_bins = 100)

# Sample cells ####
n_w <- n_h <- 40
ncells <- 0.9*n_w*n_h

bbox <- sim$search_sample(c("C" = ncells), n_w, n_h)
sim$sample_cells("SC", bbox$lower_corner, bbox$upper_corner)
plot_tissue(sim, num_of_bins = 100)

forest <- sim$get_samples_forest()
```

## Build mutation engine and add CN event

```{r, message=FALSE, warning=FALSE}
# Get genomics urls ####
reference_url <- paste0("https://ftp.ensembl.org/pub/grch37/current/",
                        "fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.",
                        "dna.chromosome.22.fa.gz")

SBS_url <- paste0("https://cancer.sanger.ac.uk/signatures/documents/2123/",
                  "COSMIC_v3.4_SBS_GRCh37.txt")

drivers_url <- paste0("https://raw.githubusercontent.com/",
                      "caravagnalab/ProCESS/main/inst/extdata/",
                      "driver_mutations_hg19.csv")

passenger_CNAs_url <- paste0("https://raw.githubusercontent.com/",
                             "caravagnalab/ProCESS/main/inst/extdata/",
                             "passenger_CNAs_hg19.csv")

germline_url <- paste0("https://www.dropbox.com/scl/fi/g9oloxkip18tr1r",
                       "m6wjve/germline_data_demo.tar.gz?rlkey=15jshul",
                       "d3bqgyfcs7fa0bzqeo&dl=1")

# build a mutation engine and place all the files in the directory "Test" ####
m_engine <- build_mutation_engine(directory = "Test",
                                  reference_src = reference_url,
                                  SBS_src = SBS_url,
                                  drivers_src = drivers_url,
                                  passenger_CNAs_src = passenger_CNAs_url,
                                  germline_src = germline_url)
```

To observe the different nature of the two events will make the deletion span a smaller segment than the amplification.

```{r}
# Define Mutants ####
# B, son of A, will have a deletion
# C, son of B, will have an amplification
# We remove passenger CNA
m_engine$add_mutant(
  mutant_name = "A",
  passenger_rates = c(SNV = 1e-7, CNA = 0),
  drivers = list(SNV("22", 10510210, "C"))
  #driver_CNAs = c(CNA(type = "A", "22", chr_pos = 10303470,len = 1e7))
)

m_engine$add_mutant("B",
  passenger_rates = c(SNV=1e-7, CNA=1e-9),
  drivers = list(SNV("22", 10510210, "C"),
                 CNA(type = "A", chr = "22", chr_pos = 20303470, len = 2e7),
                 CNA(type = "D", chr = "22", chr_pos = 20303470, len = 1e7))
)

m_engine$add_mutant(
  mutant_name = "C",
  passenger_rates = c(SNV=1e-7, CNA = 1e-9),
  drivers = list(SNV("22", 10510210, "C"))
  #driver_CNAs = c(CNA(type = "A", chr = "22", chr_pos = 10303470, len = 1e7))
)

m_engine$add_exposure(coefficients = c(SBS13 = 0.4, SBS1 = 0.6))

# Load sample forest and place 0 pre-neolpastic mutations
phylo_forest <- m_engine$place_mutations(forest, 1000)
```

## Sequencing results

Simulate the sequencing...

```{r}
# Simulate sequencing ####
seq_results <- simulate_seq(phylo_forest, coverage = 80)

seq_to_long <- function(seq_results) {
  # get names of samples
  sample_names <- strsplit(colnames(seq_results)[grepl("VAF", colnames(seq_results), fixed = TRUE)], ".VAF") %>% unlist()

  sn <- sample_names[1]
  seq_df <- lapply(sample_names, function(sn) {
    cc <- c("chr", "chr_pos", "ref", "alt", colnames(seq_results)[grepl(sn, colnames(seq_results), fixed = T)])

    seq_results[, cc] %>%
      `colnames<-`(c("chr", "chr_pos", "ref", "alt", "occurences", "coverage", "VAF")) %>%
      dplyr::mutate(sample_name = sn)
  }) %>% do.call("bind_rows", .)

  seq_df %>%
    dplyr::rename(from = chr_pos, DP = coverage, NV = occurences, ALT = alt) %>%
    dplyr::mutate(to = from) %>%
    dplyr::select(chr, from, to, ALT, NV, DP, VAF, sample_name)
}
```

```{r, warning=FALSE}
all_CNA <- phylo_forest$get_sampled_cell_CNAs() 
cna_segment <- all_CNA %>% 
  distinct(begin, end, allele, src.allele, type) %>% 
  mutate(id = paste(type, begin,end,allele,src.allele, sep = ':' )) %>% 
  select(-allele, -src.allele)

seq_df_gm <- seq_to_long(seq_results %>% dplyr::filter(classes == "germinal")) %>% filter(VAF < 0.99) 
seq_df_gm <-  seq_df_gm %>% mutate(DR = seq_df_gm$DP /  (seq_df_gm %>% filter(sample_name == 'SA') %>% pull(DP))) 

seq_df_gm <- bind_cols(seq_df_gm, cna_id = rep(NA, nrow(seq_df_gm)))
for (cna in seq(1,nrow(cna_segment))){
  seg <- cna_segment[cna,]
  seq_df_gm <- seq_df_gm %>% mutate(cna_id = ifelse(from %in% seq(seg$begin, seg$end), seg$id, cna_id))
}

germline <- seq_df_gm %>%
  ggplot(mapping = aes(x=from, y=VAF, color = cna_id)) +
  geom_point(size = 0.1) +
  ylim(0,1) +
  ylab('BAF') +
  facet_wrap(~ sample_name) + 

  seq_df_gm %>%
  ggplot(mapping = aes(x=from, y=DR, color = cna_id)) +
  geom_point(size = 0.1) +
  ylab('DR') +
  facet_wrap(~ sample_name) + 
  
seq_df_gm %>%
  ggplot(mapping = aes(x=VAF, fill = cna_id)) +
  geom_histogram(binwidth = 0.01,  position = 'identity', alpha = 0.6) +
  xlim(-0.05, 1.05) +
  xlab('BAF') +
  facet_wrap(~ sample_name) +
   
plot_layout(nrow = 3, guides = 'collect') & theme(legend.position = 'bottom') 

# Somatic stuff
seq_df_som <- seq_to_long(seq_results %>% dplyr::filter(!(classes %in% c("germinal"))))
seq_df <- bind_cols(seq_df_som, cna_id = rep(NA, nrow(seq_df_som)))
for (cna in seq(1,nrow(cna_segment))){
  seg <- cna_segment[cna,]
  seq_df <- seq_df %>% mutate(cna_id = ifelse(from %in% seq(seg$begin, seg$end), seg$id, cna_id))
}

seq_df <- seq_df %>%
  #dplyr::left_join(phylo_forest$get_sampled_cell_SNVs() %>% dplyr::rename(from = chr_pos), by = "from") %>% 
  filter(VAF >= 0.08) 

somatic <- seq_df %>%
  ggplot(mapping = aes(x=from, y=VAF, color = cna_id)) +
  geom_point(size = 0.5) +
  ylim(0, 1) +
  facet_wrap(~ sample_name) + 
  
seq_df %>%
  ggplot(mapping = aes(x=VAF, fill = cna_id)) +
  geom_histogram(binwidth = 0.01, alpha = 0.6, position = 'identity') +
  xlim(-0.05, 1.05) +
  facet_wrap(~ sample_name) +
  plot_layout(nrow = 2, guides = 'collect') & theme(legend.position = 'bottom')

germline
somatic
```
