---
title: 'ProCESS: SPN07 - part 2'
author: "Alice and Riccardo"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: united
    highlight: tango
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
```

```{r import, include=FALSE}
rm(list = ls())
setwd("~/Documents/GitHub/ProCESS-examples/SPN07/scripts")
library(ProCESS)
library(dplyr)
library(ggplot2)
library(ggpubr)
```

## Create mutation engine

```{r build_mutation_engine, results='hide', message=FALSE, warning=FALSE}
m_engine <- build_mutation_engine(setup_code = "GRCh38", context_sampling= 100)
```


```{r assign_mutation_rates, results='hide', message=FALSE, warning=FALSE}
#m_engine$get_known_drivers() %>% filter(driver_gene == 'PTEN')
m_engine$add_mutant(mutant_name = "1",
                    passenger_rates = c(SNV = 5e-9, CNA = 1e-11),
                    drivers = list(list('PTEN R130*', allele=0))
                    )

# shift = CNAqc:::get_reference('hg38') %>% filter(chr=='chr10') %>% pull(from)

m_engine$add_mutant(mutant_name = "2",
                    passenger_rates = c(SNV = 5e-9, CNA = 1e-11),
                    drivers = list(
                      CNA("D", "10", 87862638 , 109292 ,allele = 1)
                    )
                    )
m_engine$add_mutant(mutant_name = "3",
                    passenger_rates = c(SNV = 5e-9, CNA = 1e-11),
                    drivers = list(
                      list('NF1 Q1174*', allele=0)
                    )
                    )
# Convert mutation in protein coordinates to genome coordinates : https://bibliome.ai/GRCh38/gene/ATRX
m_engine$add_mutant(mutant_name = "4",
                    passenger_rates = c(SNV = 5e-9, CNA = 1e-11),
                    drivers = list(
                      SNV('X', 2719, 'T','C') # ARTX R907*
                    )
                    )
m_engine$add_mutant(mutant_name = "5",
                    passenger_rates = c(SNV = 1e-8, CNA = 1e-11),
                    drivers = list(
                      SNV('2', 1082, 'A','G') # MSH6 c.1082G>A	p.R361H
                    )
                    )
m_engine$add_mutant(mutant_name = "6",
                    passenger_rates = c(SNV = 1e-8, CNA = 1e-11),
                    drivers = list(
                      list('TP53 R248W')
                    )
                    )
```

Add exposure 
```{r add_exposures, results='hide', message=FALSE, warning=FALSE}
m_engine$add_exposure(c(SBS1 = .8, SBS5 = .2))
m_engine$add_exposure(time = 7.85,
                      c(SBS11= 1)) # Chemotherapy active from 7.85 to 10.85
m_engine$add_exposure(time = 10.85,
                      c(SBS1 = .8, SBS5 = .2))
```

Problem: is it possible to change mutation rate?

## Place Mutations

```{r place_mutations, results='hide', message=FALSE, warning=FALSE}
samples_forest <- load_samples_forest("forest_sampling_2.sff")
phylo_forest <- m_engine$place_mutations(samples_forest, 1000)
```

```{r show_mutations, results='hide', message=FALSE, warning=FALSE}
# phylo_forest$get_sampled_cell_mutations() %>% head()
# phylo_forest$get_sampled_cell_CNAs() %>% head()
# phylo_forest$get_germline_mutations() %>% head()
phylo_forest$save("phyloforest.sff")
```



## Sequencing

```{r simulate_sequencing, results='hide', message=FALSE, warning=FALSE}
phylo_forest <- load_phylogenetic_forest("phyloforest.sff")
seq_results <- simulate_seq(phylo_forest, coverage = 50)
# seq_results %>% head
```



